File: binary_tree/sum.cs

using System;

using DataStructures;

class BinaryTreeSum
{
	static void BuildBinaryTree(BinaryTree<NodeValueInt> tree)
	{
		BinaryNode<NodeValueInt> n0 = tree.root; n0.SetSatelliteData(new NodeValueInt(-1));
		
		BinaryNode<NodeValueInt> n1 = n0.CreateLeft(new NodeValueInt(2));
		BinaryNode<NodeValueInt> n2 = n0.CreateRight(new NodeValueInt(4));

		BinaryNode<NodeValueInt> n3 = n1.CreateLeft(new NodeValueInt(5));
		BinaryNode<NodeValueInt> n4 = n1.CreateRight(new NodeValueInt(6));
		BinaryNode<NodeValueInt> n5 = n2.CreateLeft(new NodeValueInt(67));
		BinaryNode<NodeValueInt> n6 = n2.CreateRight(new NodeValueInt(100));
	}

	private static void Main(string[] args)
	{
		BinaryTree<NodeValueInt> tree = new BinaryTree<NodeValueInt>(NodeValueInt.zero);
		BuildBinaryTree(tree);
		Console.WriteLine("Sum: {0}", tree.GetSum().value);
	}
}

File: binary_tree/sum.cpp

#include "../binary_tree/binary_tree.hpp"
#include <iostream>

static void BuildBinaryTree(BinaryTree<int>& tree)
{
	BinaryNode<int>* n0 = tree.getRoot(); n0->setSatelliteData(-1);

	BinaryNode<int>* n1 = n0->createLeft(2);
	BinaryNode<int>* n2 = n0->createRight(4);

	BinaryNode<int>* n3 = n1->createLeft(5);
	BinaryNode<int>* n4 = n1->createRight(6);
	BinaryNode<int>* n5 = n2->createLeft(67);
	BinaryNode<int>* n6 = n2->createRight(100);
}

int main(int argc, const char* argv[])
{
	BinaryTree<int> tree;
	BuildBinaryTree(tree);
	std::cout << "Sum: " << tree.getSum() << std::endl;


	// calls the BinaryTree<int>::operator BinaryNode<int>&() overload
	BinaryNode<int>& node = tree;
	// calls the BinaryTree<int>::operator BinaryNode<int>*() overload
	BinaryNode<int>* nodePtr = tree;
	
	return 0;
}